### ISSUE TITLE
fluent API?

### ISSUE THREAD

**Issue by tomwayson** [Esri/cedar#315]
#I know @benstoltz wants this, and I suspect that @ajturner does too (w/ suggestions like `chart.statistic('SUM(value)')`).

I've been hesitant b/c previously there was not a clear distinction between the parts of API that were doing async things vs non async, and all fluent operations have to be sync. However, now that we're starting to isolate async operations (#308 and #314), I'm open to using a fluent API for sync operations.

That said, if we're going to do this, I think we need to get rid of the properties - it's one or the other. So instead of:

```js


// the current property based API
console.log(chart.datasets);
chart.datasets[1].data = newData
chart.series[0].value = { field: newFieldName, label: newFieldLabel }
chart.render()

// vs a new fluent API
console.log(chart.datasets())
chart.series(0).value = { field: newFieldName, label: newFieldLabel }
chart.data({
    'secondDatasetsName': newData
  })
  .render()
```

I have mixed feelings about the latter, and it would be a lot of work to come up w/ a good API for v1 that will further delay it's release.

**Comment by ajturner** [comment]
I would prefer not to change to the fluent API b/c of 1) likely delay and 2) complicated DX (developer experience)

can we hold until v2?

**Comment by ajturner** [comment]
I do really like the _addressable datasets by name_. And you make a case back in [#314](https://github.com/Esri/cedar/pull/314#issuecomment-335981186) for internally why fluent is cleaner.

**Comment by tomwayson** [comment]
Thanks for weighing in @ajturner.

Re: "complicated DX", I think that cuts both ways. Besides, devs love fluent (jQuery, D3, Leaflet).

We can hold off until v2, but it's really tempting to spend a little more time after resolving [these issues](https://github.com/Esri/cedar/labels/alpha.0) to sketch out a fluent API and see if we can get it in there. At least something that we could iterate on during alpha/beta.

**Comment by benstoltz** [comment]
@ajturner  I agree that the actual implementation of this could be pushed back to v2, however I agree with @tomwayson that we should determine the naming now. 

I do like the notion of a fluent api and I'd lean more towards a leaflet style naming choice where names are more clearly understandable. For example `queryData` vs `query`

**Comment by ajturner** [comment]
@tomwayson what is a rough estimate on timeframe if we moved to fluent on v1?

**Comment by tomwayson** [comment]
@ajturner tomorrow 😄

**Comment by tomwayson** [comment]
Here's my proposal for the API:

```js
// assume arrays of datasets/series defined above
const definition = {
  type: 'bar',
  datasets,
  series
}
const c = new cedar.Chart('chartDiv', definition)

// definition methods
c.getDefinition() // this._defintion
c.setDefinition({}) // this

// working w/ datasets
c.getDatasets() // this._definition.datasets
c.setDatasets([]) // this
c.getDatasets('someName') // this._definition.datasets w/ the name = 'someName'
c.setDatasets('someName', {}) // this

// working w/ series
c.getSeries() // this._definition.series
c.setSeries([]) // this
c.getSeries(1) // this._definition.series[1]
c.setSeries(0, {}) // this

// working w/ type and specification
c.getType() // this._definition.type
c.setType('pie') // this
c.getSpecification() // this._definition.specification
c.setSpecification({}) // this

// querying data
// all of these invocations return a hash in the form of { 'datasetName': featureLayerQueryResponse }
c.queryData() // queries all datasets w/ a .url property
c.queryData('datasetName') // only query data for that dataset 

// working w/ chart (merged) data
c.getData() // this._data
// this._data should not be set directly, instead it's updated from the datasets
// this invocation would only be useful if all datasets were inline
c.updateData() // this
// this invocation is what is (currently) used after running queryData()
c.updateData(queryResponseHash) // this

// rendering the chart
c.render() // this
```

The first iteration would likely not include the overloads (so just `c.getDatasets()`, not `c.getDatasets('datasetName')`, but even w/o overloads we could replace the current properties and put the shape of the API out there to iterate on.

**Comment by tomwayson** [comment]
The above would be the "Leaflet style". Below is the "jQuery style"

```js
// definition methods
c.definition() // this._defintion
c.definition({}) // this

// working w/ datasets
c.datas() // this._definition.datasets
c.datasets([]) // this
c.datasets('someName') // this._definition.datasets w/ the name = 'someName'
c.datasets('someName', {}) // this

// working w/ series
c.series() // this._definition.series
c.series([]) // this
c.series(1) // this._definition.series[1]
c.series(0, {}) // this

// working w/ type and specification
c.type() // this._definition.type
c.type('pie') // this
c.specification() // this._definition.specification
c.specification({}) // this

// querying data
// all of these invocations return a hash in the form of { 'datasetName': featureLayerQueryResponse }
c.query() // queries all datasets w/ a .url property
c.query('datasetName') // only query data for that dataset 

// working w/ chart (merged) data
c.data() // this._data
// this._data should not be set directly, instead it's derived from the datasets
// this invocation would only be useful if all datasets were inline
c.data() // this
// this invocation is what is (currently) used after running query()
c.data(queryResponseHash) // this

// rendering the chart
c.render() // this
```

Choose. Choose and perish.

I kinda prefer the "jQuery style." Each has their sticky bits. "jQuery style" provides a clean solution to the problem of 'series' being both singular and plural, but it is a little less clear what's going on w/ `.data()`.

**Comment by benstoltz** [comment]
I too prefer the 'jquery style', with the sole exception of `data` in which case I prefer the leaflet style. Thus I have chosen and hereby do perish.

**Comment by tomwayson** [comment]
Let's go w/ jQuery style except for `c.updateData(queryResponseHash)` - that will be easy enough to alias or change later.

**Comment by tomwayson** [comment]
Giving this a shot now, the get/set overloading seems to make TS unhappy, and I don't know how to cheer it up again, so switching over to Leaflet style.

**Comment by tomwayson** [comment]
resolved in #321
